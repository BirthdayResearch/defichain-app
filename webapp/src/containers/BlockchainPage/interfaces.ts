export interface ITxn {
  hash: string;
  time: string;
  froms: {
    address: string;
    amount: number | string;
  }[];
  tos: {
    address: string;
    amount: number | string;
  }[];
  unit: string;
}

export interface IBlockData {
  nTxns: number;
  difficulty: number;
  height: number;
  bits: string;
  version: number;
  nonce: number;
  hash: string;
  merkleRoot: string;
}
