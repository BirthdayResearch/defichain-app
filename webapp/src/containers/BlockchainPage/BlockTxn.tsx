import React, { Component } from "react";
import { Button, Card } from "reactstrap";
import { MdContentCopy, MdInfo, MdArrowForward } from "react-icons/md";
import styles from "./BlockTxn.module.scss";
import CopyToClipboard from "react-copy-to-clipboard";
import classnames from "classnames";
import { I18n } from "react-redux-i18n";

interface BlockTxnProps {
  txn: {
    hash: string;
    time: string;
    froms: Array<{
      address: string;
      amount: number | string;
    }>;
    tos: Array<{
      address: String;
      amount: number | string;
    }>;
  };
}

interface BlockTxnState {
  copied: boolean;
}

class BlockTxn extends Component<BlockTxnProps, BlockTxnState> {
  state = {
    copied: false,
  };

  handleCopy = () => {
    this.setState({
      copied: true,
    });
    setTimeout(() => {
      this.setState({
        copied: false,
      });
    }, 600);
  };

  render() {
    const txn = this.props.txn;
    const froms = txn.froms;
    let fromsRender;

    if (froms.length) {
      fromsRender = froms.map((from) => (
        <div className={styles.from}>
          <span className={styles.address}>
            {from.address
              ? from.address
              : I18n.t("containers.blockChainPage.blockTxn.unparsedAddress")}
            }
          </span>
          <span className={styles.amount}>
            {from.amount} {I18n.t("containers.blockChainPage.blockTxn.dFI")}
          </span>
        </div>
      ));
    } else {
      fromsRender = (
        <div className={styles.fromsEmpty}>
          {I18n.t("containers.blockChainPage.blockTxn.noInputs")}
        </div>
      );
    }

    return (
      <Card className={`${styles.txn} styles.txn`}>
        <div className={styles.header}>
          <span className={styles.hash}>
            <div
              className={classnames(
                { "d-block": this.state.copied },
                styles.copiedIndicator
              )}
            >
              {I18n.t("containers.blockChainPage.blockTxn.copied")}
            </div>
            <span className={styles.hashSpan}>{txn.hash}</span>
            <CopyToClipboard text={txn.hash}>
              <Button
                color="link"
                size="sm"
                className="padless ml-2"
                onClick={this.handleCopy}
              >
                <MdContentCopy />
              </Button>
            </CopyToClipboard>
          </span>
          <span className={styles.time}>
            <span className={styles.timeSpan}>{txn.time}</span>
            <Button color="link" size="sm" className="padless ml-2">
              <MdInfo />
            </Button>
          </span>
        </div>
        <div className={styles.addresses}>
          <div className={styles.froms}>{fromsRender}</div>
          <div className={styles.arrow}>
            <MdArrowForward />
          </div>
          <div className={styles.tos}>
            {txn.tos.map((to) => (
              <div className={styles.to}>
                <span className={styles.address}>
                  {to.address
                    ? to.address
                    : I18n.t(
                        "containers.blockChainPage.blockTxn.unparsedAddress"
                      )}
                </span>
                <span className={styles.amount}>
                  {to.amount} {I18n.t("containers.blockChainPage.blockTxn.dFI")}
                </span>
              </div>
            ))}
          </div>
        </div>
      </Card>
    );
  }
}

export default BlockTxn;
