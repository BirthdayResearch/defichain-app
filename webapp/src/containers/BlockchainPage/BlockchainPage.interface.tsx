export interface BlockchainPageProps {}

export interface BlockchainPageState {}

export interface BlockchainTableProps {
  blocks: Array<{
    height: number;
    age: string;
    txns: string;
    minerName: string;
    minerID: number;
    size: string;
  }>;
  isLoadingBlocks: boolean;
  isBlocksLoaded: Boolean;
  blocksLoadError: string;
  fetchBlocks: Function;
}

export interface BlockchainTableState {}

export interface BlockPageProps {
  txns: Array<{
    hash: string;
    time: string;
    froms: Array<{
      address: string;
      amount: number;
    }>;
    tos: Array<{
      address: string;
      amount: number;
    }>;
  }>;
  fetchTxns: Function;
}

export interface BlockPageState {
  txns: Array<{
    hash: string;
    time: string;
    froms: Array<{
      address: string;
      amount: number | string;
    }>;
    tos: Array<{
      address: String;
      amount: number | string;
    }>;
  }>;
}

export interface BlockTxnProps {
  txn: {
    hash: string;
    time: string;
    froms: Array<{
      address: string;
      amount: number | string;
    }>;
    tos: Array<{
      address: String;
      amount: number | string;
    }>;
  };
}

export interface BlockTxnState {
  copied: boolean;
}

export interface MinerPageProps {}

export interface MinerPageState {}

export interface RouteParams {
  id?: string;
  height?: string;
}
