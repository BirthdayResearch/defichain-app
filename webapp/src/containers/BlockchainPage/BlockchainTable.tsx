import React, { Component } from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import {
  Card,
  Table,
  Pagination,
  PaginationItem,
  PaginationLink,
} from "reactstrap";
import {
  MdChevronLeft,
  MdChevronRight,
  MdFirstPage,
  MdLastPage,
} from "react-icons/md";
import styles from "./BlockchainTable.module.scss";
import {
  BlockchainTableProps,
  BlockchainTableState,
} from "./BlockchainPage.interface";
import { I18n } from "react-redux-i18n";
import { fetchBlocksRequest } from "./reducer";

class BlockchainTable extends Component<
  BlockchainTableProps,
  BlockchainTableState
> {
  componentDidMount() {
    this.props.fetchBlocks();
  }

  render() {
    return (
      <>
        <Card className={styles.card}>
          <div className={`${styles.tableResponsive} table-responsive-xl`}>
            <Table className={styles.table}>
              <thead>
                <tr>
                  <th>
                    {I18n.t("containers.blockChainPage.blockChainTable.height")}
                  </th>
                  <th>
                    {I18n.t("containers.blockChainPage.blockChainTable.age")}
                  </th>
                  <th>
                    {I18n.t(
                      "containers.blockChainPage.blockChainTable.transactions"
                    )}
                  </th>
                  <th>
                    {I18n.t(
                      "containers.blockChainPage.blockChainTable.minedBy"
                    )}
                  </th>
                  <th>
                    {I18n.t("containers.blockChainPage.blockChainTable.size")}
                  </th>
                </tr>
              </thead>
              <tbody>
                {this.props.blocks.map((block) => (
                  <tr key={block.height}>
                    <td>
                      <Link to={`/blockchain/block/${block.height}`}>
                        {block.height}
                      </Link>
                    </td>
                    <td>
                      <div>{block.age}</div>
                    </td>
                    <td>
                      <div>{block.txns}</div>
                    </td>
                    <td>
                      <div>
                        <Link to={`/blockchain/miner/${block.minerID}`}>
                          {block.minerName}
                        </Link>
                      </div>
                    </td>
                    <td>
                      <div>{block.size}</div>
                    </td>
                  </tr>
                ))}
              </tbody>
            </Table>
          </div>
        </Card>
        <div className="d-flex justify-content-between align-items-center mt-3">
          <div>{I18n.t("containers.blockChainPage.blockChainTable.count")}</div>
          <Pagination className={styles.pagination}>
            <PaginationItem>
              <PaginationLink first href="#">
                <MdFirstPage />
              </PaginationLink>
            </PaginationItem>
            <PaginationItem>
              <PaginationLink previous href="#">
                <MdChevronLeft />
              </PaginationLink>
            </PaginationItem>
            <PaginationItem>
              <PaginationLink href="#">1</PaginationLink>
            </PaginationItem>
            <PaginationItem>
              <PaginationLink href="#">2</PaginationLink>
            </PaginationItem>
            <PaginationItem>
              <PaginationLink href="#">3</PaginationLink>
            </PaginationItem>
            <PaginationItem>
              <PaginationLink next href="#">
                <MdChevronRight />
              </PaginationLink>
            </PaginationItem>
            <PaginationItem>
              <PaginationLink last href="#">
                <MdLastPage />
              </PaginationLink>
            </PaginationItem>
          </Pagination>
        </div>
      </>
    );
  }
}

const mapStateToProps = (state) => {
  const {
    blocks,
    isBlocksLoaded,
    isLoadingBlocks,
    blocksLoadError,
  } = state.blockchain;
  return {
    blocks,
    isBlocksLoaded,
    isLoadingBlocks,
    blocksLoadError,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    fetchBlocks: () => dispatch(fetchBlocksRequest()),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(BlockchainTable);
