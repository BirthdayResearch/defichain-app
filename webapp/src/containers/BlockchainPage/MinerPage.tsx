import React, { Component } from "react";
import { Helmet } from "react-helmet";
import { Button } from "reactstrap";
import { MdArrowBack } from "react-icons/md";
import { NavLink, RouteComponentProps } from "react-router-dom";
import { I18n } from "react-redux-i18n";

interface MinerPageProps {}

interface MinerPageState {}

interface RouteParams {
  id?: string;
  height?: string;
}

class MinerPage extends Component<
  MinerPageProps & RouteComponentProps<RouteParams>,
  MinerPageState
> {
  render() {
    return (
      <div className="main-wrapper">
        <Helmet>
          <title>
            {I18n.t("containers.blockChainPage.minerPage.miner")}
            {`${this.props.match.params.id} â€“`}
            {I18n.t("containers.blockChainPage.minerPage.defiBlockChainClient")}
          </title>
        </Helmet>
        <header className="header-bar">
          <Button
            to="/blockchain"
            tag={NavLink}
            color="link"
            className="header-bar-back"
          >
            <MdArrowBack />
            <span className="d-lg-inline">
              {I18n.t("containers.blockChainPage.minerPage.blockchain")}
            </span>
          </Button>
          <h1>
            {I18n.t("containers.blockChainPage.minerPage.miner")}&nbsp;
            {this.props.match.params.id}
          </h1>
        </header>
        <div className="content">
          <section>
            {I18n.t("containers.blockChainPage.minerPage.miner")}
          </section>
        </div>
      </div>
    );
  }
}

export default MinerPage;
