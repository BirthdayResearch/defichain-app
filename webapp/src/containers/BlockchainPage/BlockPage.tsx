import React, { Component } from 'react';
import { Helmet } from 'react-helmet';
import { connect } from 'react-redux';
import { Button, Row, Col } from 'reactstrap';
import { MdArrowBack, MdChevronLeft, MdChevronRight } from 'react-icons/md';
import { NavLink, RouteComponentProps } from 'react-router-dom';
import KeyValueLi from '../../components/KeyValueLi/KeyValueLi';
import BlockTxn from './BlockTxn';
import { I18n } from 'react-redux-i18n';
import { fetchTxnsRequest } from './reducer';
import {
  BLOCKCHAIN_BASE_PATH,
  BLOCKCHAIN_BLOCK_BASE_PATH,
} from '../../constants';

interface Txns {
  hash: string;
  time: string;
  froms: {
    address: string;
    amount: number | string;
  }[];
  tos: {
    address: string;
    amount: number | string;
  }[];
}

interface BlockPageProps {
  txns: Txns[];
  fetchTxns: () => void;
}

interface BlockPageState {
  txns: Txns[];
}

interface RouteParams {
  id?: string;
  height?: string;
}

class BlockPage extends Component<
  BlockPageProps & RouteComponentProps<RouteParams>,
  BlockPageState
> {
  componentDidMount = () => {
    this.props.fetchTxns();
  };

  render() {
    const { height } = this.props.match.params;
    return (
      <div className='main-wrapper'>
        <Helmet>
          <title>
            {I18n.t('containers.blockChainPage.blockPage.title', {
              blockNo: height,
            })}
          </title>
        </Helmet>
        <header className='header-bar'>
          <Button
            to={BLOCKCHAIN_BASE_PATH}
            tag={NavLink}
            color='link'
            className='header-bar-back'
          >
            <MdArrowBack />
            <span className='d-lg-inline'>
              {I18n.t('containers.blockChainPage.blockPage.blockchain')}
            </span>
          </Button>
          <h1>
            {I18n.t('containers.blockChainPage.blockPage.block')}&nbsp;
            {height}
          </h1>
        </header>
        <div className='content'>
          <section className='mb-5'>
            <Row className='mb-4'>
              <Col md='6'>
                <KeyValueLi
                  label={I18n.t(
                    'containers.blockChainPage.blockPage.noOfTransactions'
                  )}
                  value='2851'
                />
              </Col>
              <Col md='6'>
                <KeyValueLi
                  label={I18n.t(
                    'containers.blockChainPage.blockPage.difficulty'
                  )}
                  value='13798783827516.416'
                />
              </Col>
              <Col md='6'>
                <KeyValueLi
                  label={I18n.t('containers.blockChainPage.blockPage.height')}
                  value={height}
                />
              </Col>
              <Col md='6'>
                <KeyValueLi
                  label={I18n.t('containers.blockChainPage.blockPage.bits')}
                  value='171465f2'
                />
              </Col>
              <Col md='6'>
                <KeyValueLi
                  label={I18n.t(
                    'containers.blockChainPage.blockPage.blockReward'
                  )}
                  value='12.5 DFI'
                />
              </Col>
              <Col md='6'>
                <KeyValueLi
                  label={I18n.t('containers.blockChainPage.blockPage.version')}
                  value='21073676288'
                />
              </Col>
              <Col md='6'>
                <KeyValueLi
                  label={I18n.t('containers.blockChainPage.blockPage.minedBy')}
                  value='Miner A'
                />
              </Col>
              <Col md='6'>
                <KeyValueLi
                  label={I18n.t('containers.blockChainPage.blockPage.nonce')}
                  value='353942907'
                />
              </Col>
              <Col>
                <KeyValueLi
                  label={I18n.t(
                    'containers.blockChainPage.blockPage.blockHash'
                  )}
                  value='00000000000000000003e1bee7555dd5ecfb2d54eaca0650d426e10f640b7f89'
                  copyable='true'
                />
              </Col>
              <Col>
                <KeyValueLi
                  label={I18n.t(
                    'containers.blockChainPage.blockPage.merkleRoot'
                  )}
                  value='2c5030ae78f6201d76f20baf77a3e3'
                  copyable='true'
                />
              </Col>
            </Row>
            <div className='d-flex justify-content-between'>
              <Button
                to={`${BLOCKCHAIN_BLOCK_BASE_PATH}/${Number.parseInt(
                  height!,
                  10
                ) - 1}`}
                tag={NavLink}
                color='outline-primary'
                className='header-bar-back'
              >
                <MdChevronLeft />
                <span className='d-lg-inline'>
                  {Number.parseInt(height!, 10) - 1}
                </span>
              </Button>
              <Button
                to={`${BLOCKCHAIN_BLOCK_BASE_PATH}/${Number.parseInt(
                  height!,
                  10
                ) + 1}`}
                tag={NavLink}
                color='outline-primary'
                className='header-bar-back'
              >
                <span className='d-lg-inline'>
                  {Number.parseInt(height!, 10) + 1}
                </span>
                <MdChevronRight />
              </Button>
            </div>
          </section>
          <section>
            <h2>
              {I18n.t('containers.blockChainPage.blockPage.transactions')}
            </h2>
            <div>
              {this.props.txns.map(txn => (
                <BlockTxn txn={txn} />
              ))}
            </div>
          </section>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  const { txns, isTxnsLoaded, isLoadingTxns, TxnsLoadError } = state.blockchain;
  return {
    txns,
    isTxnsLoaded,
    isLoadingTxns,
    TxnsLoadError,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    fetchTxns: () => dispatch(fetchTxnsRequest()),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(BlockPage);
