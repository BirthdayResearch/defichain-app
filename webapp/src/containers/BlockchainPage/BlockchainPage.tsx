import React, { Component } from "react";
import { Helmet } from "react-helmet";
import BlockchainTable from "./BlockchainTable";
import { I18n } from "react-redux-i18n";

interface BlockchainPageProps {}

interface BlockchainPageState {}

class BlockchainPage extends Component<
  BlockchainPageProps,
  BlockchainPageState
> {
  render() {
    return (
      <div className="main-wrapper">
        <Helmet>
          <title>
            {I18n.t("containers.blockChainPage.blockChainPage.title")}
          </title>
        </Helmet>
        <header className="header-bar">
          <h1>
            {I18n.t("containers.blockChainPage.blockChainPage.blockchain")}
          </h1>
        </header>
        <div className="content">
          <section className="mb-0">
            <BlockchainTable />
          </section>
        </div>
      </div>
    );
  }
}

export default BlockchainPage;
