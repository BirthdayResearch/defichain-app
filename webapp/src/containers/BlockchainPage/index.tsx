import React from 'react';
import { Helmet } from 'react-helmet';
import BlockchainTable from './components/BlockchainTable';
import { I18n } from 'react-redux-i18n';
import Header from '../HeaderComponent';
import { MdLaunch } from 'react-icons/md';
import { ButtonGroup, Button } from 'reactstrap';
import openNewTab from '../../utils/openNewTab';
import {
  DEFICHAIN_BLOCKS,
  MAIN,
} from '../../constants';
import { getNetworkType, getPageTitle } from '../../utils/utility';

const BlockchainPage = () => {
  const explorerLink =
    getNetworkType() === MAIN ? `${DEFICHAIN_BLOCKS}` : `${DEFICHAIN_BLOCKS}?network=TestNet`;
  return (
    <div className='main-wrapper'>
      <Helmet>
        <title>
          {getPageTitle(
            I18n.t('containers.blockChainPage.blockChainPage.title')
          )}
        </title>
      </Helmet>
      <Header>
        <h1>{I18n.t('containers.blockChainPage.blockChainPage.blockchain')}</h1>
        <ButtonGroup>
          <Button color='link' onClick={() => openNewTab(explorerLink)}>
            <MdLaunch />
            <span className='d-lg-inline'>
              {I18n.t('containers.blockChainPage.blockChainPage.explorer')}
            </span>
          </Button>
        </ButtonGroup>
      </Header>
      <div className='content'>
        <section className='mb-0'>
          <BlockchainTable />
        </section>
      </div>
    </div>
  );
};

export default BlockchainPage;
