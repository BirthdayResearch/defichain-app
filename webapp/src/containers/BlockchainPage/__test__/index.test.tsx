import React from 'react';
import BlockchainPage from '../index';
import { shallow } from 'enzyme';

describe('BlockChainPage component', () => {
  it('should check for snapshot', () => {
    const wrapper = shallow(<BlockchainPage />);
    expect(wrapper).toMatchSnapshot();
  });
});
