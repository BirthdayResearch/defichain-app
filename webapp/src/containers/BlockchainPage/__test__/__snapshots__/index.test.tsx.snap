// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockChainPage component should check for snapshot 1`] = `
<div
  className="main-wrapper"
>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
      Blockchain - DeFi Wallet
    </title>
  </HelmetWrapper>
  <Connect(Header)>
    <h1>
      Blockchain
    </h1>
    <ButtonGroup
      role="group"
      tag="div"
    >
      <Button
        color="link"
        onClick={[Function]}
        tag="button"
      >
        <MdLaunch />
        <span
          className="d-lg-inline"
        >
          Explorer
        </span>
      </Button>
    </ButtonGroup>
  </Connect(Header)>
  <div
    className="content"
  >
    <section
      className="mb-0"
    >
      <Connect(BlockchainTable) />
    </section>
  </div>
</div>
`;
